test-services/test-probe/src/context.rs

```diff
@@ -0,0 +1,72 @@
+//! Define routes for exposing the contents of the application's
+//! rendering context.
+
+use crate::{config,
+            error::Result};
+use actix_web::{web,
+                HttpRequest,
+                HttpResponse};
+use serde_json;
+use std::{fs,
+          path::{Components,
+                 Path,
+                 PathBuf}};
+
+/// Show render context rooted at `path`
+pub fn show(config: web::Data<config::Config>,
```

Comment:
I think this could be simplified like:
```rust
pub fn show(config: web::Data<config::Config>,
            path: web::Path<PathBuf>)
            -> actix_web::Result<HttpResponse> {
    let json =
        extract_data(&config.render_context_file, path.into_inner()).map_err(actix_web::error::ErrorBadRequest)?;
    ...
}
```

Reply:
Ah, neat! I think with the pointer-related changes in 29c091c, I'm not sure if this helps, but definitely worth remembering for the future.

Reply:
I think it would still be useful. You could pull out the pointer in the function signature with `pointer: web::Path<String>` this would avoid the need to unwrap the `Option`. It also has better static guarentees as `actix-web` will not even route a request here if it does not match the function signature.

Reply:
Hrm... right now, I think I _do_ want it to route here if there's no path provided, in order to deal with the "show the entire context" use case.

Reply:
That is a good point. If you want to allow showing the full context for both `/context` and `/context/` you would have to wrap the path in an `Option`. It does work but is not an obvious improvement.

```rust
/// Show render context rooted at `path`
#[allow(clippy::needless_pass_by_value)] // Required by actix-web API
pub fn show(config: web::Data<config::Config>,
            pointer: Option<web::Path<String>>)
            -> actix_web::Result<HttpResponse> {
    // A JSON pointer of "" means "everything". Pointers starting with
    // "/" describe paths into the structure. We'll chop off any
    // trailing "/" since that will end up looking for an item with a
    // key of "", which we're just not going to deal with.
    let pointer = pointer.map(|p| p.into_inner());
    let pointer = match pointer.as_ref().map(String::as_str) {
        None | Some("") => Cow::Borrowed(""),
        Some(path) => Cow::Owned(format!("/{}", path.trim_end_matches("/"))),
    };
    let json = read_json(&config.render_context_file).map_err(actix_web::error::ErrorBadRequest)?;
    if let Some(value) = json.pointer(&pointer) {
        Ok(HttpResponse::Ok().json(value))
    } else {
        Ok(HttpResponse::NotFound().finish())
    }
}
```